### DISCUSSION TITLE
How do I enable debug logging?

### DISCUSSION THREAD

**Discussion by danieltroger** [solidjs/solid#628] [Q&A] [answered]
Hi, I'm trying to get solid to run in IE11 because a colleague of mine wrote something in solid and left and now it's my job to get this finished and deployed, and we need IE11 support.

Before I start rewriting without solid, I thought giving making it run in IE11 a try would be a good idea.

I have managed to transpile solid and brought a boatload of polyfills - it even runs successfully in firefox 48 now - but it just silently fails in IE11.

So I wanted to ask, how do I enable debug logging of solid so that I can see where things go wrong?

**Comment by ryansolid** [comment, marked as answer]
There is a dev mode. We have dev outputs you can alias. One way is through node export conditions but failing that alias solid-js, solid-js/web, and solid-js/store to "solid-js/dist/dev.js", "solid-js/web/dist/dev.js", "solid-js/store/dist/dev.js"

**Reply by danieltroger** [reply]
Thanks @ryansolid. I ran `yarn unplug` and moved the files. Just to be curious, is there an easier way to alias it? I'm using yarn berry.

Also, this check throws in older browsers (`ReferenceError: globalThis is not defined`):
```
if (globalThis) {
    if (!globalThis.Solid$$) globalThis.Solid$$ = true;
    else console.warn("You appear to have multiple instances of Solid. This can lead to unexpected behavior.");
}
```

You need to either do `if (window.globalThis) {` or `if (typeof globalThis === "object" && globalThis) {`

Otherwise the check is useless, or am I stupid?

Also why can't I have multiple instances of solid? As a third party javascript developer that's a bummer. What if the site my widget is for already has solid?

**Reply by danieltroger** [reply]
Yoooooooooooooo @ryansolid I got solid fucking working in IE11!!!!! I thought it'd be impossible because you're using proxies but apparently you're not hitting the cases where the proxy polyfill fails ???? I've patched parcel to transpile it and added a boatload of polyfills - it silently failed without Math.imul polyfill but now in dev mode it printed the error and I added it and it fucking works

**Comment by atk** [comment]
It's not impossible with transpilation, merely impossible with polyfills only.

**Reply by danieltroger** [reply]
Well of course it's impossible to run modern javascript syntax like arrow functions in IE11 without transpiling it. But are you talking about transpiling anything more than what babel usually does? Maybe I only got it partially working if that's the case.

**Reply by atk** [reply]
Depends on your babel config. Usually, you specify the features you want babel to emulate through transpiling and polyfills through presets and/or browser versions.

**Reply by danieltroger** [reply]
With "what babel usually does" I meant what's a feature of babel. For example, it's impossible to polyfill the proxying of an array in it's entirety. It would be possible to work around that by transpiling the code that changes the array to also call the proxy trap function.
This would be super advanced and does not exist afaik. Babel doesn't do this - usually, you could patch babel or install plugins that maybe do this.

So what I wanted to know is if you could hypothetically make solid work in IE11 by transpiling it (with hypothetical polyfills) or if you *actually* can do this with existing tooling. It seems like the latter is the case because if you meant the former you'd have understood my question and the component which I got running in IE11 with existing transpilation works perfectly fine.

**Reply by ryansolid** [reply]
I think proxies and async/await functions are not really polyfillable. I know there are transpile handlers for async/await, but proxies I'm less clear about how they go about and do that.

**Reply by atk** [reply]
You can certainly transpile solid. However, if you tried to use the transpiled solid with non-transpiled JS using certain features, you'd hit some limits.

**Reply by danieltroger** [reply]
@ryansolid yeah. Async/await can be transpiled without any drawbacks. It works without problems. (You can even use transpiled await on non transpiled promises and vise-versa @atk, it works by the common denominator of everything just being a `.then`-able object). There is no transpilation for `Proxy` that I'm aware of, but there is this polyfill by google: https://github.com/GoogleChrome/proxy-polyfill

I thought that all change detection in solid is based on `Proxy` but that doesn't seem to be the case (after reading your excellent 50-line reactive library code). So that might explain why it works.
Regarding the `Proxy` polyfill, it basically works on everything named. So i.e. in `{ key: 'value' }` you could intercept get and set on `key` but if someone would add or remove another key you couldn't intercept it. The same goes for arrays. Intercepting functions with `new Proxy(fn, {apply: (target, this_arg, argument_list) => {}});` works fine. Also it freezes objects sometimes (don't remember the exact cases) so if you'd for example try to proxy `localStorage` it would freeze localStorage for the whole site lol.

Hopes this helps somehow. But I think claiming compatibility could help with this library's adoption - I do imagine some looking at the readme page and seeing `The last 2 versions of modern evergreen browsers and Node LTS.` and thinking "hmm, neat for my personal website but it doesn't sound production ready for my business". If you have any questions how I polyfilled or transpiled, feel free to ask.

**Reply by ryansolid** [reply]
Solid does use proxies for things like spreads so and helpers around managing props even if you don't get into stores. So Imagine there are things that may not work neatly. I've seen that polyfill and it might work, but there are scenarios were we do listen to things they don't intercept like ownProperty etc.. I wasn't going to take my chances on saying we support something and then not fully support it and leave people with that missed expectation. I'd rather say we don't and if it happens to work for someone that's cool.
